src/lazyfree.c

```diff
@@ -212,3 +220,13 @@ void freeTrackingRadixTreeAsync(rax *tracking) {
     atomicIncr(lazyfree_objects,tracking->numele);
     bioCreateLazyFreeJob(lazyFreeTrackingTable,1,tracking);
 }
+
+/* Free lua_scripts dict, if the dict is huge enough, free it in async way. */
+void freeLuaScriptsAsync(dict *lua_scripts) {
+    if (dictSize(lua_scripts) > LAZYFREE_THRESHOLD) {
```

Comment:
i don't think this condition is sufficient.

in all other places a lazy free is dependent on a user input or a config, and this threshold just defines if it worth while to send the job to the other thread or keep it in the main thread just because the list is short.

i.e the user first needs to decide if he's willing to accept the async freeing side effects (not atomic), or not.

so to be consistent with that, we should either add a check for `lazyfree_lazy_server_del` or add an `ASYNC` argument for `SCRIPT FLUSH`

Reply:
Agree with you, I think both of your suggestions should be taken:
1. Provide `SCRIPT FLUSH ASYNC` for users to actively trigger.
2. `SCRIPT FLUSH` judges whether to release asynchronously according to `lazyfree_lazy_server_del`.

Reply:
on a second thought, i'm actually not sure `lazyfree-lazy-server-del` is the right config, maybe it should be `lazyfree-lazy-user-del` or neither.
looking at what FLUSHB does, it should be neither.

Reply:
I don't think we should use lazyfree-lazy-server-del or lazyfree-lazy-user-del, since they are already defined and don't mean this. 

I also don't think we should be adding ASYNC into command arguments. It's not really something you should be deciding on a case by case basis, and it adds overhead to client developers to support and us to document.

I think we should introduce a general new flag for lazyfree-policy, that covers every other lazy free usage not covered by an existing flag. I think it would even be nice to update the default value for the other lazy-free configs to add a third option, use-default, which would use the lazyfree-policy value unless an explicit value was set.

Reply:
> I think we should introduce a general new flag for lazyfree-policy, that covers every other lazy free usage not covered by an existing flag.

I agree with your suggestion of adding a new flag, for example: `lazyfree-lazy-server-free`. However, it is more difficult for us to find the remaining code that is not covered by lazyfree. The problem of `scriptingRelease` is encountered in our production environment, but there are many places where `dictRelease` is called, see [here](https:// github.com/redis/redis/search?q=dictRelease) . In addition to `dictRelease`, there are `listRelease`, `raxFree`, etc.

> I think it would even be nice to update the default value for the other lazy-free configs to add a third option, use-default, which would use the lazyfree-policy value unless an explicit value was set.

Here I have a little doubt, for example: you mean that `lazyfree-lazy-eviction` will have three optional values: `yes`,`no`,`use-default`?   If use-default is set, then the rest of the lazyfree configuration is this value?

Reply:
Should we just use "lazyfree-lazy-server-free" and have it cover everything else? I just don't want us to keep adding individual flags for every little thing.

Reply:
@madolson @oranagra Because `lazyfree-lazy-server-del` or `lazyfree-lazy-user-del` already has its own meaning.

Do you agree with me to introduce a new configuration `lazyfree-lazy-server-free` to control the remaining places that are not lazyfree? For example, we mentioned `scriptRelease()` above, but there may be other free areas that need to be controlled in the future.

The default value of `lazyfree-lazy-server-free` is no.

Reply:
@yangbodong22011 Hey, the redis maintainers synced offline and have a concrete recommendation. Sorry for the bit of churn. 

We decided we're fine just adding more flags as we go along, so just have a new flag called something like ```lazyfree-lazy-scriptcache-flush``` that controls the behavior of calling "SCRIPT FLUSH" with no arguments.

We also suggest that we support both "SCRIPT FLUSH [ASYNC|SYNC]" to force it one way or another. So, to summarize:
1. "SCRIPT FLUSH": Uses the value from from the config
2. "SCRIPT FLUSH ASYNC": Always flushes the config in an async manner.
3. "SCRIPT FLUSH SYNC": Always flushes the config in a sync manner.

Reply:
@madolson Thank you for your reply, and sorry to disturb you if you are on vacation. I agree and will implement it in accordance with your comments.

Reply:
@madolson maybe we have a misunderstanding or my memory misleads me, i'm not sure that's what i understood we decided.
I was under the impression that we're gonna put SCRIPT FLUSH and FLUSHDB in the same bucket.
give both a SYNC and ASYNC argument, and have some global config that sets the default behavior when neither argument is provided.

Let me try to think this over, i'll split the configs to 3 groups
1. `lazyfree-lazy-eviction`, `lazyfree-lazy-expire`, `lazyfree-lazy-server-del`, `replica-lazy-flush`:
these are all about deletions that the server does implicitly, not as a result of a specific command.
2. `lazyfree-lazy-user-del` and DEL / UNLINK:
in this case there's a specific command for async, but since it's widely used we added a server config to change the legacy command behavior.
3. FLUSHDB / FLUSHALL / SCRIPT FLUSH: 
These have (or should have) an ASYNC and SYNC arguments, and a single config defining the default behavior when no argument is specified. i.e. `lazyfree-lazy-user-flush` which has a default of `no` for backwards compatibility.

looking at the delta between your last post and mine, i guess the differences are:
1. in mine there's a single config that controls both SCRIPT FLUSH and FLUSHALL.
2. in yours you only listed a new config that's specific for SCRIPT FLUSH.

so i'm also ok with taking the one you suggested `lazy-scriptcache-flush`, but for symmetry i think we should then also add `lazy-db-flush` config, and i think we were trying to come up with a single config that controls several aspects rather than add one for each.

Reply:
@oranagra The reason why you provide the `lazy-db-flush` parameter for flushall/flushdb is to be compatible with the user's old code only to call flushall/flushdb, but still control lazyfree through the server, right?

>  i think we were trying to come up with a single config that controls several aspects rather than add one for each.

How about `lazyfree-lazy-user-flush`?

```
# When FLUSHDB,FLUSHALL,SCRIPT FLUSH with no arguments, it will decide whether
# lazyfree is required according to the following configuration.
lazyfree-lazy-user-flush no
```

Reply:
update code, PR title and PR comments.

the current behavior is as follows:

- FLUSH[ALL|DB],SCRIPT FLUSH: Determine sync or async according to the value of lazyfree-lazy-user-flush.
- FLUSH[ALL|DB],SCRIPT FLUSH ASYNC: Always flushes the cache in an async manner.
- FLUSH[ALL|DB],SCRIPT FLUSH SYNC: Always flushes the config in a sync manner.

Reply:
I think this falls in line with the spirit of what we discussed, so I'm fine with it.

Reply:
Hi guys. I have some questions: 
- `freeLuaScriptsAsync` This function name metions the behavior will be a async way. But we still calculate the dict size. And if dictsize not reach `LAZYFREE_THRESHOLD 64` . It will become a sync way. In other ways, `scriptingRelease` require a `async` arg to call `freeLuaScriptsAsync`. But inside `freeLuaScriptsAsync`, it could be a sync way
- so do `freeObjAsync` function
- and also the comment of  `freeTrackingRadixTreeAsync` seems wrong

There is not problem with those questions. Just my thinking

Reply:
@enjoy-binbin Yes, `LAZYFREE_THRESHOLD` controls whether it is really async free.

About `freeTrackingRadixTreeAsync` comment, it's wrong, it will direct async free and do not judge object's size.
```
/* Free an object, if the object is huge enough, free it in async way. */
void freeTrackingRadixTreeAsync(rax *tracking) {
    atomicIncr(lazyfree_objects,tracking->numele);
    bioCreateLazyFreeJob(lazyFreeTrackingTable,1,tracking);
}
```
and another problem is we should use `tracking->numnodes`(Every macro node in the Stream is one allocation) instead of `tracking->numele`.

Now we have two points to fix:
1, use `tracking->numnodes` instead of `tracking->numele`
2. Judge size to async free or fix comments.
@oranagra @madolson WDYT?

Reply:
Received a reply so soon. cool. I am willing to correct it if needed

And the mean time. Do we need to be consistent before and after method behavior?

Reply:
freeXxxxAsync doesn't necessarily means it'll be released by the thread.
it means it will be potentially async (which has implication on memory being reclaimed in time or not), but it doesn't mean it'll always be done by the thread.
even the UNLINK function doesn't always release asynchronously, it considers that and makes a efficiency decision (it will be inefficient to do it async if you just need to release a few pointers).
so i don't see a problem about the initial complain on `freeLuaScriptsAsync`, or maybe i didn't understand it.

i agree that `freeTrackingRadixTreeAsync` and also `freeSlotsToKeysMapAsync` need to be extended with a check for LAZYFREE_THRESHOLD (against `numnodes`, not `numele`).
please note that not every rax structure that we release this way needs to look at `numnodes`. if the rax was holding any points in it's elements that need to be released too, then we should look at `numele` or `nuele+numnodes`, but in this case but these rax structures are holding only keys (no values), so it only needs to release the nodes.

regarding `lazyfree_objects`, it is a statistics metric, not a real counter of released allocations. it seems to denote the number of objects / elements released, not actual allocations.
However, when we release a list object (composed of a linked list of ziplists), we increment it only by one (not the number of elements in the list, and not the number of linked-list nodes), so for from that point of view, we can increment it only by one here too. (if we do that, we need to update the call to `atomicDecr` too).
Again, just a metric, and i don't have any major argument for incrementing by either [1, numele, numnodes].

Reply:
`freeXxxxAsync doesn't necessarily means it'll be released by the thread.`
I misunderstood....
I make pr https://github.com/redis/redis/pull/8969 to extend `freeTrackingRadixTreeAsync` and `freeSlotsToKeysMapAsync `
About `lazyfree_objects`, according your says. I did not change it 

Thanks a lot for your reply. Very detailed. @oranagra